Fix sdkServer.logLevel nil in Gameserver in local-mode

govargo opened the pull request:
**What this PR does / Why we need it**:

/kind bug

In local-mode sdk-server, the `SdkServer.LogLevel` is always ignored, even if the users set the LogLevel.
It always set to the info level.
In the current codebase, the local-sdkserver's behavier is defined by `pkg/sdkserver/localsdk.go`.

In the `pkg/sdkserver/localsdk.go`, there is no function for setting logLevel.
So the LogLevel is always set to the Info(Level=4).

In this PR, I added the function which set logLevel to the local-sdkserver if specified and added the unit test for this.

**After this PR**

User can set LogLevel

```
LogLevel: Debug
{"ctlConf":{"Address":"localhost","IsLocal":true,"LocalFile":"/Users/xxxxxx/git/agones/examples/simple-game-server/gameserver.yaml","Delay":0,"Timeout":0,"Test":"","TestSdkName":"test","GRPCPort":9357,"HTTPPort":9358},"featureGates":"CustomFasSyncInterval=false\u0026Example=true\u0026NodeExternalDNS=true\u0026PlayerAllocationFilter=false\u0026PlayerTracking=false\u0026SDKGracefulTermination=false\u0026StateAllocationFilter=false","message":"Starting sdk sidecar","severity":"info","source":"main","time":"2022-05-14T17:32:34.496763+09:00","version":"dev"}
{"filePath":"/Users/xxxxxx/git/agones/examples/simple-game-server/gameserver.yaml","message":"Reading GameServer configuration","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2022-05-14T17:32:34.497066+09:00"}
{"grpcEndpoint":"localhost:9357","message":"Starting SDKServer grpc service...","severity":"info","source":"main","time":"2022-05-14T17:32:34.501695+09:00"}
{"httpEndpoint":"localhost:9358","message":"Starting SDKServer grpc-gateway...","severity":"info","source":"main","time":"2022-05-14T17:32:34.502205+09:00"}
{"message":"Ready request has been received!","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2022-05-14T17:32:39.863248+09:00"}
{"message":"Received ready request","severity":"debug","source":"*sdkserver.LocalSDKServer","time":"2022-05-14T17:32:39.863334+09:00"}
```

```
LogLevel: Error
{"ctlConf":{"Address":"localhost","IsLocal":true,"LocalFile":"/Users/xxxxxx/git/agones/examples/simple-game-server/gameserver.yaml","Delay":0,"Timeout":0,"Test":"","TestSdkName":"test","GRPCPort":9357,"HTTPPort":9358},"featureGates":"CustomFasSyncInterval=false\u0026Example=true\u0026NodeExternalDNS=true\u0026PlayerAllocationFilter=false\u0026PlayerTracking=false\u0026SDKGracefulTermination=false\u0026StateAllocationFilter=false","message":"Starting sdk sidecar","severity":"info","source":"main","time":"2022-05-14T17:31:12.769861+09:00","version":"dev"}
{"filePath":"/Users/xxxxxx/git/agones/examples/simple-game-server/gameserver.yaml","message":"Reading GameServer configuration","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2022-05-14T17:31:12.770279+09:00"}
```

**Which issue(s) this PR fixes**:

Closes #2221

**Special notes for your reviewer**:

In this PR, I updated the proto file and go file about `sdk.proto`.
If other further updates are required, please let me know.

Thank you

govargo commented:
/assign @alekser

markmandel commented on proto/sdk/sdk.proto:
```diff
@@ -129,13 +129,20 @@ message GameServer {
             int32 initial_delay_seconds = 4;
         }
+
+        message SdkServer {
```
Not sure I'm understanding why the proto change? I don't see any major reason to pass log level information to the SDK user?

govargo replied:
This `SdkServer` and `SdkServer.logLevel` are not used in following section in my PR.
https://github.com/googleforgames/agones/pull/2580/files#diff-a02544ece56edb873e82b56146e058a2cc36f010391d705b826899d17db824ffR666-R679

But these are used in test codes which I added.
https://github.com/googleforgames/agones/pull/2580/files#diff-a11e2ffffc7affd5f7777c02cba93464d213f1389e2a5451edf8ad317deec07eR114-R137

In the current proto file, `SdkServer` is not defined.
```
	fixture := &agonesv1.GameServer{
		ObjectMeta: metav1.ObjectMeta{Name: "stuff"},
		Spec: agonesv1.GameServerSpec{
			SdkServer: agonesv1.SdkServer{LogLevel: "Debug"},
		},
	}
```
Without `SdkServer` definition, the above code returns the nil spec GameServer.

markmandel replied:
🤔 you have access to the private variables within LocalSDK, so you could check the `*gs` value directly, rather than loading it this way - then you don't need the proto change.

govargo replied:
Thank you for your comment. 
As you said, there was no need to change the proto.
It could confirm just by checking the private variables in LocalSDK.

I reverted proto changes.

govargo commented:
```
Step #24 - "sdk-conformance": {"message":"shutting down sdk server","severity":"info","source":"main","time":"2022-05-24T09:14:52.550647681Z"}
Step #24 - "sdk-conformance": {"message":"Could not find a request which was expected","request":"shutdown","sdkName":"cpp","severity":"error","source":"*sdkserver.LocalSDKServer","time":"2022-05-24T09:14:52.55074598Z"}
Step #24 - "sdk-conformance": {"expected":["ready","allocate","setlabel","setannotation","gameserver","health","shutdown","watch","reserve"],"message":"Testing Failed","received":["health","watch","ready","allocate","reserve","gameserver","setlabel","setannotation","ready"],"sdkName":"cpp","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2022-05-24T09:14:52.550766394Z"}
```

It looks like that the test of cpp failed because the assertion order was messed up. `["health","watch","ready","allocate","reserve","gameserver","setlabel","setannotation","ready"]`

govargo commented:
I changed the `TIMEOUT` from 30s to 40s in `build/includes/sdk.mk`.

I observed sometimes that the localsdk server immediately shut down before test finishes, due to timeout(30s).
This is why I changed the timeout setting. 

Please let me know, if this is not desirable.

```
Step #24 - "sdk-conformance": {"message":"shutting down sdk server","severity":"info","source":"main","time":"2022-05-24T09:14:52.550647681Z"}
Step #24 - "sdk-conformance": {"message":"Could not find a request which was expected","request":"shutdown","sdkName":"cpp","severity":"error","source":"*sdkserver.LocalSDKServer","time":"2022-05-24T09:14:52.55074598Z"}
Step #24 - "sdk-conformance": {"expected":["ready","allocate","setlabel","setannotation","gameserver","health","shutdown","watch","reserve"],"message":"Testing Failed","received":["health","watch","ready","allocate","reserve","gameserver","setlabel","setannotation","ready"],"sdkName":"cpp","severity":"info","source":"*sdkserver.LocalSDKServer","time":"2022-05-24T09:14:52.550766394Z"}
Step #24 - "sdk-conformance": Shutting down after 10 seconds...
Step #24 - "sdk-conformance": Could not run Shutdown():failed to connect to all addresses. Exiting!
Step #24 - "sdk-conformance": terminate called without an active exception
Step #24 - "sdk-conformance": /root/sdktest.sh: line 18:     8 Aborted                 (core dumped)
```

govargo commented:
The test failure https://github.com/googleforgames/agones/pull/2580#issuecomment-1135636130 is same as #2346.
(I found out later)

markmandel reviewed (approved):
LGTM.

Let's see if the extra timeout helps with the sdk flakiness. 🤞🏻
